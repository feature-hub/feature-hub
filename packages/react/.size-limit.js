// @ts-check

module.exports = [
  {
    path: 'lib/esm/index.js',
    limit: '3.1 KB',
  },
];
