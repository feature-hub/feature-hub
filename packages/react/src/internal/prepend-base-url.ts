import urljoin from 'url-join';

function isAbsolute(url: string): boolean {
  return /^https?:\/\//.test(url);
}

/**
 * Prepends the given `base` - which can be relative or absolute - to the `url`,
 * unless the `url` is absolute.
 */
export function prependBaseUrl(base: string | undefined, url: string): string {
  if (!base || isAbsolute(url)) {
    return url;
  }

  return urljoin(base, url);
}
