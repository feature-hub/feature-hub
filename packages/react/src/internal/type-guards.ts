import {
  DomFeatureApp,
  FeatureApp,
  ReactFeatureApp,
} from '../feature-app-container';

export function isDomFeatureApp(
  featureApp: object
): featureApp is DomFeatureApp {
  return typeof (featureApp as DomFeatureApp).attachTo === 'function';
}

export function isReactFeatureApp(
  featureApp: object
): featureApp is ReactFeatureApp {
  return typeof (featureApp as ReactFeatureApp).render === 'function';
}

export function isFeatureApp(featureApp: unknown): featureApp is FeatureApp {
  if (typeof featureApp !== 'object' || !featureApp) {
    return false;
  }

  return isDomFeatureApp(featureApp) || isReactFeatureApp(featureApp);
}
