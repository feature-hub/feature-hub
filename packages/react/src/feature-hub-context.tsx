import {AsyncSsrManagerV1} from '@feature-hub/async-ssr-manager';
import {FeatureAppManager, Logger} from '@feature-hub/core';
import * as React from 'react';

type SomeRequired<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> &
  Required<Pick<T, K>>;

export interface Css {
  readonly href: string;
  readonly media?: string;
}

export interface FeatureHubContextProviderValue {
  /**
   * The `FeatureAppManager` singleton instance.
   */
  readonly featureAppManager: FeatureAppManager;

  /**
   * A custom logger that shall be used instead of `console`.
   */
  readonly logger?: Logger;

  /**
   * The Async SSR Manager Feature Service that is bound to the integrator. It
   * is only provided on the server.
   */
  readonly asyncSsrManager?: AsyncSsrManagerV1;

  /**
   * A callback that the integrator provides on the server, mainly for the
   * [[FeatureAppLoader]], to add client URLs of those Feature Apps that are
   * rendered on the server, so that they can be preloaded on the client before
   * hydration. This method might be called multiple times with the same URL
   * during server-side rendering. Therefore deduplication should be considered.
   *
   * @param url The **client** URL of a Feature App that is rendered on the
   * server.
   *
   * @param moduleType The **client** module type of the Feature App that is
   * rendered on the server. This value can be used by the provided
   * [[FeatureAppManagerOptions.moduleLoader]].
   */
  addUrlForHydration?(url: string, moduleType?: string): void;

  /**
   * A callback that the integrator provides on the server, mainly for the
   * [[FeatureAppLoader]], to add stylesheets for those Feature Apps that are
   * rendered on the server, so that they can be added to the document before
   * being sent to the client. This method might be called multiple times with
   * the same list of stylesheets (per Feature App) during server-side
   * rendering. Therefore deduplication should be considered.
   *
   * @param stylesheets A list of stylesheets for a Feature App that is rendered
   * on the server.
   */
  addStylesheetsForSsr?(stylesheets: Css[]): void;
}

export type FeatureHubContextConsumerValue = SomeRequired<
  FeatureHubContextProviderValue,
  'logger'
>;

const noFeatureHubContextValueErrorMessage =
  'No Feature Hub context was provided! There are two possible causes: 1.) No FeatureHubContextProvider was rendered in the React tree. 2.) A Feature App that renders itself a FeatureAppLoader or a FeatureAppContainer did not declare @feature-hub/react as an external package.';

export const FeatureHubContext = React.createContext<
  FeatureHubContextProviderValue | undefined
>(undefined);

/**
 * The integrator must render a `FeatureHubContextProvider` to provide the
 * `FeatureAppManager` singleton instance, as well as the Async SSR Manager
 * Feature Service (optionally), to Feature Hub context consumers, e.g. the
 * [[FeatureAppLoader]] or [[FeatureAppContainer]].
 */
export const FeatureHubContextProvider = FeatureHubContext.Provider;

/**
 * A `FeatureHubContextConsumer` can be used to retrieve the `FeatureAppManager`
 * singleton instance, as well as the Async SSR Manager Feature Service (if it
 * is provided by the integrator). Its main purpose is to be used inside the
 * [[FeatureAppLoader]] and [[FeatureAppContainer]].
 */
export function FeatureHubContextConsumer(
  props: React.ConsumerProps<FeatureHubContextConsumerValue>
): JSX.Element {
  return (
    <FeatureHubContext.Consumer>
      {(featureHubContextValue) => {
        if (!featureHubContextValue) {
          throw new Error(noFeatureHubContextValueErrorMessage);
        }

        // Provide `console` as the default logger.
        const {logger = console, ...rest} = featureHubContextValue;

        return props.children({logger, ...rest});
      }}
    </FeatureHubContext.Consumer>
  );
}
