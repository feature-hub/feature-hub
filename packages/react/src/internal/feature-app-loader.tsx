import {AsyncSsrManagerV0} from '@feature-hub/async-ssr-manager';
import {FeatureAppDefinition, FeatureAppManagerLike} from '@feature-hub/core';
import * as React from 'react';
import {FeatureAppContainer} from '../feature-app-container';

export interface Css {
  readonly href: string;
  readonly media?: string;
}

export interface InternalFeatureAppLoaderProps {
  readonly featureAppManager: FeatureAppManagerLike;
  readonly src: string;
  readonly serverSrc?: string;
  readonly css?: Css[];
  readonly idSpecifier?: string;
  readonly asyncSsrManager?: AsyncSsrManagerV0;
}

interface InternalFeatureAppLoaderState {
  readonly featureAppDefinition?: FeatureAppDefinition<unknown>;
  readonly hasError?: boolean;
}

const inBrowser =
  typeof window === 'object' &&
  typeof document === 'object' &&
  document.nodeType === 9;

export class InternalFeatureAppLoader extends React.PureComponent<
  InternalFeatureAppLoaderProps,
  InternalFeatureAppLoaderState
> {
  public readonly state: InternalFeatureAppLoaderState = {};

  private errorReported = false;
  private mounted = false;

  public constructor(props: InternalFeatureAppLoaderProps) {
    super(props);

    const {
      featureAppManager,
      src: browserSrc,
      serverSrc,
      asyncSsrManager
    } = props;

    const src = inBrowser ? browserSrc : serverSrc;

    if (!src) {
      if (inBrowser) {
        throw new Error('No src provided.');
      }

      return;
    }

    const asyncFeatureAppDefinition = featureAppManager.getAsyncFeatureAppDefinition(
      src
    );

    if (asyncFeatureAppDefinition.error) {
      this.reportError(asyncFeatureAppDefinition.error);

      if (!inBrowser) {
        throw asyncFeatureAppDefinition.error;
      }

      this.state = {hasError: true};
    } else if (asyncFeatureAppDefinition.value) {
      this.state = {featureAppDefinition: asyncFeatureAppDefinition.value};
    } else if (!inBrowser && asyncSsrManager) {
      asyncSsrManager.rerenderAfter(asyncFeatureAppDefinition.promise);
    }
  }

  public async componentDidMount(): Promise<void> {
    this.mounted = true;

    this.appendCss();

    if (this.state.featureAppDefinition) {
      return;
    }

    const {featureAppManager, src} = this.props;

    const asyncFeatureAppDefinition = featureAppManager.getAsyncFeatureAppDefinition(
      src
    );

    try {
      const featureAppDefinition = await asyncFeatureAppDefinition.promise;

      if (this.mounted) {
        this.setState({featureAppDefinition});
      }
    } catch (error) {
      this.reportError(error);

      if (this.mounted) {
        this.setState({hasError: true});
      }
    }
  }

  public componentWillUnmount(): void {
    this.mounted = false;
  }

  public render(): React.ReactNode {
    const {featureAppDefinition, hasError} = this.state;

    if (hasError) {
      // An error UI could be rendered here.
      return null;
    }

    if (!featureAppDefinition) {
      // A loading UI could be rendered here.
      return null;
    }

    return (
      <FeatureAppContainer
        featureAppDefinition={featureAppDefinition}
        idSpecifier={this.props.idSpecifier}
      />
    );
  }

  private appendCss(): void {
    if (!this.props.css) {
      return;
    }

    for (const {href, media} of this.props.css) {
      if (!document.querySelector(`link[href="${href}"]`) && document.head) {
        const attributes = {rel: 'stylesheet', href, media: media || 'all'};
        const element = Object.assign(
          document.createElement('link'),
          attributes
        );

        document.head.appendChild(element);
      }
    }
  }

  private reportError(error: Error): void {
    if (this.errorReported) {
      return;
    }

    this.errorReported = true;

    const {idSpecifier, src: browserSrc, serverSrc} = this.props;
    const src = inBrowser ? browserSrc : serverSrc;

    console.error(
      idSpecifier
        ? `The Feature App for the src ${JSON.stringify(
            src
          )} and the ID specifier ${JSON.stringify(
            idSpecifier
          )} could not be rendered.`
        : `The Feature App for the src ${JSON.stringify(
            src
          )} could not be rendered.`,
      error
    );
  }
}
