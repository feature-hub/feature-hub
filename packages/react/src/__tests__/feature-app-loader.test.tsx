// tslint:disable:no-implicit-dependencies

import {AsyncSsrManagerV1} from '@feature-hub/async-ssr-manager';
import {
  AsyncValue,
  FeatureAppDefinition,
  FeatureAppManager,
} from '@feature-hub/core';
import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import {CustomFeatureAppRenderingParams, FeatureAppLoader} from '..';
import {FeatureHubContextProvider} from '../feature-hub-context';
import {InternalFeatureAppContainer} from '../internal/internal-feature-app-container';
import {logger} from './logger';
import {TestErrorBoundary} from './test-error-boundary';

interface MockAsyncSsrManager extends AsyncSsrManagerV1 {
  scheduleRerender: ((promise: Promise<unknown>) => void) & jest.Mock;
}

jest.mock('../internal/internal-feature-app-container', () => ({
  InternalFeatureAppContainer: jest.fn(
    () => 'mocked InternalFeatureAppContainer'
  ),
}));

describe('FeatureAppLoader', () => {
  let mockFeatureAppManager: FeatureAppManager;
  let mockGetAsyncFeatureAppDefinition: jest.Mock;
  let mockAsyncFeatureAppDefinition: AsyncValue<FeatureAppDefinition<unknown>>;
  let mockAsyncSsrManager: MockAsyncSsrManager;
  let mockAddUrlForHydration: jest.Mock;
  let mockAddStylesheetsForSsr: jest.Mock;
  let consoleErrorSpy: jest.SpyInstance;

  const usingTestErrorBoundaryConsoleErrorCalls = [
    [
      expect.stringContaining(
        'React will try to recreate this component tree from scratch using the error boundary you provided, TestErrorBoundary.'
      ),
    ],
  ];

  const noErrorBoundaryConsoleErrorCalls = [
    [
      expect.stringContaining(
        'Consider adding an error boundary to your tree to customize error handling behavior.'
      ),
    ],
  ];

  const expectConsoleErrorCalls = (expectedConsoleErrorCalls: unknown[][]) => {
    try {
      expect(consoleErrorSpy.mock.calls).toEqual(expectedConsoleErrorCalls);
    } finally {
      consoleErrorSpy.mockClear();
    }
  };

  beforeEach(() => {
    consoleErrorSpy = jest.spyOn(console, 'error');

    if (document.head) {
      document.head.innerHTML = '';
    }

    mockAsyncFeatureAppDefinition = new AsyncValue(
      new Promise<FeatureAppDefinition<unknown>>(jest.fn())
    );

    mockGetAsyncFeatureAppDefinition = jest.fn(
      () => mockAsyncFeatureAppDefinition
    );

    mockFeatureAppManager = ({
      getAsyncFeatureAppDefinition: mockGetAsyncFeatureAppDefinition,
      createFeatureAppScope: jest.fn(),
      preloadFeatureApp: jest.fn(),
    } as Partial<FeatureAppManager>) as FeatureAppManager;

    mockAsyncSsrManager = {
      scheduleRerender: jest.fn(),
      renderUntilCompleted: jest.fn(),
    };

    mockAddUrlForHydration = jest.fn();
    mockAddStylesheetsForSsr = jest.fn();
  });

  afterEach(() => {
    expect(consoleErrorSpy).not.toHaveBeenCalled();
    consoleErrorSpy.mockRestore();
  });

  it('throws an error when rendered without a FeatureHubContextProvider', () => {
    expect(() =>
      TestRenderer.create(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      )
    ).toThrowError(
      new Error(
        'No Feature Hub context was provided! There are two possible causes: 1.) No FeatureHubContextProvider was rendered in the React tree. 2.) A Feature App that renders itself a FeatureAppLoader or a FeatureAppContainer did not declare @feature-hub/react as an external package.'
      )
    );

    expectConsoleErrorCalls(noErrorBoundaryConsoleErrorCalls);
  });

  const renderWithFeatureHubContext = (
    node: React.ReactNode,
    options: {customLogger?: boolean; handleError?: (error: Error) => void} = {}
  ) => {
    const {customLogger = true, handleError = jest.fn()} = options;

    return TestRenderer.create(
      <FeatureHubContextProvider
        value={{
          featureAppManager: mockFeatureAppManager,
          asyncSsrManager: mockAsyncSsrManager,
          addUrlForHydration: mockAddUrlForHydration,
          addStylesheetsForSsr: mockAddStylesheetsForSsr,
          logger: customLogger ? logger : undefined,
        }}
      >
        <TestErrorBoundary handleError={handleError}>{node}</TestErrorBoundary>
      </FeatureHubContextProvider>
    );
  };

  it('throws an error if no src is provided', () => {
    const handleError = jest.fn();

    const testRenderer = renderWithFeatureHubContext(
      <FeatureAppLoader featureAppId="testId" src="" />,
      {handleError}
    );

    expect(handleError.mock.calls).toEqual([[new Error('No src provided.')]]);
    expect(testRenderer.toJSON()).toBe('test error boundary');
    expectConsoleErrorCalls(usingTestErrorBoundaryConsoleErrorCalls);
  });

  describe('when rendering initially', () => {
    describe('when given no children function', () => {
      it('renders nothing', () => {
        const testRenderer = renderWithFeatureHubContext(
          <FeatureAppLoader featureAppId="testId" src="example.js" />
        );

        expect(testRenderer.toJSON()).toBeNull();
      });
    });
  });

  describe('without a css prop', () => {
    it('does not change the document head', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      );

      expect(document.head).toMatchSnapshot();
    });

    it('does not try to add stylesheets for SSR', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      );

      expect(mockAddStylesheetsForSsr).not.toHaveBeenCalled();
    });
  });

  describe('with a css prop', () => {
    it('appends link elements to the document head', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          css={[
            {href: 'https://example.com/foo.css'},
            {href: 'bar.css', media: 'print'},
          ]}
        />
      );

      expect(document.head).toMatchSnapshot();
    });

    it('does not add the stylesheets for SSR', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          css={[{href: 'foo.css'}]}
        />
      );

      expect(mockAddStylesheetsForSsr).not.toHaveBeenCalled();
    });

    describe('when the css has already been appended', () => {
      it('does not append the css a second time', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            css={[{href: 'foo.css'}]}
          />
        );

        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            css={[{href: 'foo.css'}]}
          />
        );

        expect(document.head).toMatchSnapshot();
      });
    });

    describe('and a baseUrl', () => {
      it('appends link elements to the document head', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            baseUrl="http://feature-hub.io"
            src="example.js"
            css={[
              {href: 'https://example.com/foo.css'},
              {href: 'bar.css', media: 'print'},
            ]}
          />
        );

        expect(document.head).toMatchSnapshot();
      });
    });
  });

  describe('when a Feature App definition is synchronously available', () => {
    let mockFeatureAppDefinition: FeatureAppDefinition<unknown>;

    beforeEach(() => {
      mockFeatureAppDefinition = {create: jest.fn()};

      mockAsyncFeatureAppDefinition = new AsyncValue(
        Promise.resolve(mockFeatureAppDefinition)
      );
    });

    it('calls getAsyncFeatureAppDefinition with the given src exactly once', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      );

      expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
        ['example.js', undefined],
      ]);
    });

    describe('with a baseUrl and a relative src', () => {
      it('calls getAsyncFeatureAppDefinition with a prepended src', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            baseUrl="/base"
            src="example.js"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['/base/example.js', undefined],
        ]);
      });
    });

    describe('with a baseUrl and an absolute src', () => {
      it('calls getAsyncFeatureAppDefinition with the absolute src', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            baseUrl="/base"
            src="http://example.com/foo.js"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['http://example.com/foo.js', undefined],
        ]);
      });
    });

    describe('with a moduleType', () => {
      it('calls getAsyncFeatureAppDefinition with the given src and moduleType', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            moduleType="a"
            serverModuleType="b"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['example.js', 'a'],
        ]);
      });
    });

    it('renders an InternalFeatureAppContainer', () => {
      const onError = jest.fn();
      const beforeCreate = jest.fn();
      const done = jest.fn();
      const children = jest.fn();

      const testRenderer = renderWithFeatureHubContext(
        <FeatureAppLoader
          src="example.js"
          featureAppId="testId"
          config="testConfig"
          beforeCreate={beforeCreate}
          done={done}
          onError={onError}
          baseUrl="/base"
          children={children}
        />
      );

      expect(testRenderer.toJSON()).toBe('mocked InternalFeatureAppContainer');

      const expectedProps = {
        baseUrl: '/base',
        beforeCreate,
        done,
        config: 'testConfig',
        featureAppDefinition: mockFeatureAppDefinition,
        featureAppId: 'testId',
        onError,
        children,
        featureAppManager: mockFeatureAppManager,
        logger,
      };

      expect(InternalFeatureAppContainer).toHaveBeenCalledWith(
        expectedProps,
        {}
      );
    });

    it('does not schedule a rerender on the Async SSR Manager', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          serverSrc="example-node.js"
        />
      );

      expect(mockAsyncSsrManager.scheduleRerender).not.toHaveBeenCalled();
    });

    it('does not add a URL for hydration', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          serverSrc="example-node.js"
        />
      );

      expect(mockAddUrlForHydration).not.toHaveBeenCalled();
    });
  });

  describe('when the async Feature App definition synchronously has an error', () => {
    let mockError: Error;

    beforeEach(() => {
      mockError = new Error('Failed to load Feature App module.');

      mockAsyncFeatureAppDefinition = new AsyncValue(
        Promise.reject(mockError),
        undefined,
        mockError
      );
    });

    it('renders nothing and logs an error (only once)', async () => {
      const testRenderer = renderWithFeatureHubContext(
        <FeatureAppLoader src="example.js" featureAppId="testId" />
      );

      try {
        await mockAsyncFeatureAppDefinition.promise;
      } catch (error) {
        expect(error).toBe(mockError);
      } finally {
        expect(testRenderer.toJSON()).toBeNull();

        expect(logger.error.mock.calls).toEqual([
          [
            'The Feature App for the src "example.js" and the ID "testId" could not be rendered.',
            mockError,
          ],
        ]);
      }
    });

    describe('with onError provided', () => {
      it('calls onError with the error', () => {
        const onError = jest.fn();

        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            onError={onError}
          />
        );

        expect(onError.mock.calls).toEqual([[mockError]]);
      });

      it('does not log the error', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            onError={jest.fn()}
          />
        );

        expect(logger.error).not.toHaveBeenCalled();
      });

      describe('when onError throws an error', () => {
        let onErrorMockError: Error;

        beforeEach(() => {
          onErrorMockError = new Error('Throwing in onError.');
        });

        it('throws the error in render', () => {
          const handleError = jest.fn();

          const testRenderer = renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              onError={() => {
                throw onErrorMockError;
              }}
            />,
            {handleError}
          );

          expect(handleError.mock.calls).toEqual([[onErrorMockError]]);
          expect(testRenderer.toJSON()).toBe('test error boundary');
          expectConsoleErrorCalls(usingTestErrorBoundaryConsoleErrorCalls);
        });
      });
    });

    describe('with children function provided', () => {
      it('calls children with error and loading=false', () => {
        const children = jest.fn().mockReturnValue(null);

        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            children={children}
          />
        );

        const expectedParameter: CustomFeatureAppRenderingParams = {
          featureAppNode: undefined,
          error: mockError,
          loading: false,
        };

        expect(children.mock.calls).toEqual([[expectedParameter]]);
      });

      it('renders what children returns', () => {
        const children = jest.fn(() => 'Custom Error UI');

        const testRenderer = renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            children={children}
          />
        );

        expect(testRenderer.toJSON()).toBe('Custom Error UI');
      });

      describe('when children function throws an error', () => {
        let childrenMockError: Error;

        beforeEach(() => {
          childrenMockError = new Error('Throwing in children.');
        });

        it('throws the error in render', () => {
          const handleError = jest.fn();

          const testRenderer = renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              children={() => {
                throw childrenMockError;
              }}
            />,
            {handleError}
          );

          expect(handleError.mock.calls).toEqual([[childrenMockError]]);
          expect(testRenderer.toJSON()).toBe('test error boundary');
          expectConsoleErrorCalls(usingTestErrorBoundaryConsoleErrorCalls);
        });
      });

      describe('when onError is also provided', () => {
        it('calls onError with the error', () => {
          const onError = jest.fn();

          renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              onError={onError}
              children={jest.fn().mockReturnValue(null)}
            />
          );

          expect(onError.mock.calls).toEqual([[mockError]]);
        });

        describe('and throws an error', () => {
          let onErrorMockError: Error;

          beforeEach(() => {
            onErrorMockError = new Error('Throwing in onError.');
          });

          it('does not call children and throws the error in render', () => {
            const handleError = jest.fn();
            const children = jest.fn();

            const testRenderer = renderWithFeatureHubContext(
              <FeatureAppLoader
                featureAppId="testId"
                src="example.js"
                onError={() => {
                  throw onErrorMockError;
                }}
                children={children}
              />,
              {handleError}
            );

            expect(children).not.toHaveBeenCalled();
            expect(handleError.mock.calls).toEqual([[onErrorMockError]]);
            expect(testRenderer.toJSON()).toBe('test error boundary');
            expectConsoleErrorCalls(usingTestErrorBoundaryConsoleErrorCalls);
          });
        });
      });
    });
  });

  describe('when a Feature App definition is loaded asynchronously', () => {
    let mockFeatureAppDefinition: FeatureAppDefinition<unknown>;

    beforeEach(() => {
      mockFeatureAppDefinition = {create: jest.fn()};

      mockAsyncFeatureAppDefinition = new AsyncValue(
        new Promise<FeatureAppDefinition<unknown>>((resolve) =>
          // defer to next event loop turn to guarantee asynchronism
          setImmediate(() => resolve(mockFeatureAppDefinition))
        )
      );
    });

    it('calls getAsyncFeatureAppDefinition with the given src exactly twice', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      );

      expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
        ['example.js', undefined],
        ['example.js', undefined],
      ]);
    });

    describe('with a baseUrl and a relative src', () => {
      it('calls getAsyncFeatureAppDefinition with a prepended src', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            baseUrl="/base"
            src="example.js"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['/base/example.js', undefined],
          ['/base/example.js', undefined],
        ]);
      });
    });

    describe('with a baseUrl and an absolute src', () => {
      it('calls getAsyncFeatureAppDefinition with the absolute src', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            baseUrl="/base"
            src="http://example.com/foo.js"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['http://example.com/foo.js', undefined],
          ['http://example.com/foo.js', undefined],
        ]);
      });
    });

    describe('with a moduleType prop', () => {
      it('calls getAsyncFeatureAppDefinition with the given src and moduleType', () => {
        renderWithFeatureHubContext(
          <FeatureAppLoader
            featureAppId="testId"
            src="example.js"
            moduleType="a"
            serverModuleType="b"
          />
        );

        expect(mockGetAsyncFeatureAppDefinition.mock.calls).toEqual([
          ['example.js', 'a'],
          ['example.js', 'a'],
        ]);
      });
    });

    it('initially renders nothing', () => {
      const testRenderer = renderWithFeatureHubContext(
        <FeatureAppLoader featureAppId="testId" src="example.js" />
      );

      expect(testRenderer.toJSON()).toBeNull();
    });

    it('renders a FeatureAppContainer when loaded', async () => {
      const testRenderer = renderWithFeatureHubContext(
        <FeatureAppLoader src="example.js" featureAppId="testId" />
      );

      await mockAsyncFeatureAppDefinition.promise;

      expect(testRenderer.toJSON()).toBe('mocked InternalFeatureAppContainer');
    });

    it('does not schedule a rerender on the Async SSR Manager', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          serverSrc="example-node.js"
        />
      );

      expect(mockAsyncSsrManager.scheduleRerender).not.toHaveBeenCalled();
    });

    it('does not add a URL for hydration', () => {
      renderWithFeatureHubContext(
        <FeatureAppLoader
          featureAppId="testId"
          src="example.js"
          serverSrc="example-node.js"
        />
      );

      expect(mockAddUrlForHydration).not.toHaveBeenCalled();
    });

    describe('when unmounted before loading has finished', () => {
      it('renders nothing', async () => {
        const testRenderer = renderWithFeatureHubContext(
          <FeatureAppLoader featureAppId="testId" src="example.js" />
        );

        testRenderer.unmount();

        await mockAsyncFeatureAppDefinition.promise;

        expect(testRenderer.toJSON()).toBeNull();
      });
    });
  });

  describe('when a Feature App definition fails to load asynchronously', () => {
    let mockError: Error;

    beforeEach(() => {
      mockError = new Error('Failed to load Feature App module.');

      mockAsyncFeatureAppDefinition = new AsyncValue(
        new Promise<FeatureAppDefinition<unknown>>((_, reject) =>
          // defer to next event loop turn to guarantee asynchronism
          setImmediate(() => reject(mockError))
        )
      );
    });

    it('renders nothing and logs an error', async () => {
      const testRenderer = renderWithFeatureHubContext(
        <FeatureAppLoader src="example.js" featureAppId="testId" />
      );

      expect.assertions(4); // Note: one of the assertions is in afterEach.

      try {
        await mockAsyncFeatureAppDefinition.promise;
      } catch (error) {
        expect(error).toBe(mockError);
      }

      expect(testRenderer.toJSON()).toBeNull();

      expect(logger.error.mock.calls).toEqual([
        [
          'The Feature App for the src "example.js" and the ID "testId" could not be rendered.',
          mockError,
        ],
      ]);
    });

    describe('with onError provided', () => {
      it('calls onError with the error', async () => {
        const onError = jest.fn();

        try {
          renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              onError={onError}
            />
          );

          await mockAsyncFeatureAppDefinition.promise;
        } catch {}

        expect(onError.mock.calls).toEqual([[mockError]]);
      });

      it('does not log the error', async () => {
        try {
          renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              onError={jest.fn()}
            />
          );

          await mockAsyncFeatureAppDefinition.promise;
        } catch {}

        expect(logger.error).not.toHaveBeenCalled();
      });

      describe('when onError throws an error', () => {
        let onErrorMockError: Error;

        beforeEach(() => {
          onErrorMockError = new Error('Throwing in onError.');
        });

        it('throws the error in render', async () => {
          const handleError = jest.fn();

          const testRenderer = renderWithFeatureHubContext(
            <FeatureAppLoader
              featureAppId="testId"
              src="example.js"
              onError={() => {
                throw onErrorMockError;
              }}
            />,
            {handleError}
          );

          try {
            await mockAsyncFeatureAppDefinition.promise;
          } catch {}

          expect(handleError.mock.calls).toEqual([[onErrorMockError]]);
          expect(testRenderer.toJSON()).toBe('test error boundary');
          expectConsoleErrorCalls(usingTestErrorBoundaryConsoleErrorCalls);
        });

        describe('when unmounted before loading has finished', () => {
          it('calls onError with the error', async () => {
            const onError = jest.fn(() => {
              throw onErrorMockError;
            });

            const testRenderer = renderWithFeatureHubContext(
              <FeatureAppLoader
                featureAppId="testId"
                src="example.js"
                onError={onError}
              />
            );

            testRenderer.unmount();

            try {
              await mockAsyncFeatureAppDefinition.promise;
            } catch {}

            expect(onError.mock.calls).toEqual([[mockError]]);
          });

          it('renders nothing', async () => {
            const testRenderer = renderWithFeatureHubContext(
              <FeatureAppLoader
                featureAppId="testId"
                src="example.js"
                onError={() => {
                  throw onErrorMockError;
                }}
              />
            );

            testRenderer.unmount();

            try {
              await mockAsyncFeatureAppDefinition.promise;
            } catch {}

            expect(testRenderer.toJSON()).toBeNull();
          });
        });
      });
    });

    describe('when unmounted before loading has finished', () => {
      it('logs an error', async () => {
        const testRenderer = renderWithFeatureHubContext(
          <FeatureAppLoader featureAppId="testId" src="example.js" />
        );

        testRenderer.unmount();

        expect.assertions(3); // Note: one of the assertions is in afterEach.

        try {
          await mockAsyncFeatureAppDefinition.promise;
        } catch (error) {
          expect(error).toBe(mockError);
        }

        expect(logger.error.mock.calls).toEqual([
          [
            'The Feature App for the src "example.js" and the ID "testId" could not be rendered.',
            mockError,
          ],
        ]);
      });
    });
  });

  describe('without a custom logger', () => {
    it('logs messages using the console', () => {
      const mockError = new Error('Failed to load Feature App module.');

      mockAsyncFeatureAppDefinition = new AsyncValue(
        Promise.reject(mockError),
        undefined,
        mockError
      );

      renderWithFeatureHubContext(
        <FeatureAppLoader src="example.js" featureAppId="testId" />,
        {customLogger: false}
      );

      expectConsoleErrorCalls([
        [
          'The Feature App for the src "example.js" and the ID "testId" could not be rendered.',
          mockError,
        ],
      ]);
    });
  });
});
