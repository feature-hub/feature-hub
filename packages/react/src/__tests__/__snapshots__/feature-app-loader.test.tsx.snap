// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureAppLoader with a css prop and a baseUrl appends link elements to the document head 1`] = `
<head>
  <link
    href="https://example.com/foo.css"
    media="all"
    rel="stylesheet"
  />
  <link
    href="http://feature-hub.io/bar.css"
    media="print"
    rel="stylesheet"
  />
</head>
`;

exports[`FeatureAppLoader with a css prop appends link elements to the document head 1`] = `
<head>
  <link
    href="https://example.com/foo.css"
    media="all"
    rel="stylesheet"
  />
  <link
    href="bar.css"
    media="print"
    rel="stylesheet"
  />
</head>
`;

exports[`FeatureAppLoader with a css prop when the css has already been appended does not append the css a second time 1`] = `
<head>
  <link
    href="foo.css"
    media="all"
    rel="stylesheet"
  />
</head>
`;

exports[`FeatureAppLoader without a css prop does not change the document head 1`] = `<head />`;
